pkg/util/servicecidr/servicecidr.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package servicecidr
+
+import (
+	"fmt"
+	"net"
+	"net/netip"
+
+	networkingv1alpha1 "k8s.io/api/networking/v1alpha1"
+	"k8s.io/apimachinery/pkg/labels"
+	networkinglisters "k8s.io/client-go/listers/networking/v1alpha1"
+)
+
+// ContainsAddress return the list of ServiceCIDR that contains the address passed as argument
+func ContainsAddress(serviceCIDRLister networkinglisters.ServiceCIDRLister, address netip.Addr) []*networkingv1alpha1.ServiceCIDR {
+	result := []*networkingv1alpha1.ServiceCIDR{}
+	serviceCIDRList, err := serviceCIDRLister.List(labels.Everything())
+	if err != nil {
+		return result
+	}
+
+	for _, serviceCIDR := range serviceCIDRList {
+		for _, cidr := range serviceCIDR.Spec.CIDRs {
+			if prefix, err := netip.ParsePrefix(cidr); err == nil { // it can not fail since is already validated
+				if PrefixContainIP(prefix, address) {
+					result = append(result, serviceCIDR)
+				}
+			}
+		}
+	}
+	return result
+}
+
+// PrefixContainIP returns true if the given IP is contained with the prefix,
+// is not the network address and also, if IPv4, is not the broadcast address.
+// This is required because the Kubernetes allocators reserve these addresses
+// so IPAddresses can not block deletion of this ranges.
+func PrefixContainIP(prefix netip.Prefix, ip netip.Addr) bool {
+	// if the IP is the network address is not contained
+	if prefix.Masked().Addr() == ip {
+		return false
+	}
+	// the broadcast address is not considered contained for IPv4
+	if !ip.Is6() {
+		ipLast, err := BroadcastAddress(prefix)
+		if err != nil || ipLast == ip {
+			return false
+		}
+	}
+	return prefix.Contains(ip)
+}
+
+// BroadcastAddress returns the broadcast address of the subnet
+// The broadcast address is obtained by setting all the host bits
+// in a subnet to 1.
+// network 192.168.0.0/24 : subnet bits 24 host bits 32 - 24 = 8
+// broadcast address 192.168.0.255
+func BroadcastAddress(subnet netip.Prefix) (netip.Addr, error) {
```

Comment:
(likewise, does not need to be exported)

Reply:
done